File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/ticket/EventCard.java
package de.tum.in.tumcampusapp.component.ui.ticket;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.jetbrains.annotations.NotNull;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.ui.overview.CardManager;
import de.tum.in.tumcampusapp.component.ui.overview.card.Card;
import de.tum.in.tumcampusapp.component.ui.overview.card.CardViewHolder;
import de.tum.in.tumcampusapp.component.ui.ticket.activity.EventDetailsActivity;
import de.tum.in.tumcampusapp.component.ui.ticket.adapter.EventsAdapter;
import de.tum.in.tumcampusapp.component.ui.ticket.model.Event;

public class EventCard extends Card {

    private Event event;

    public EventCard(Context context) {
        super(CardManager.CARD_EVENT, context, "card_event", false);
    }

    public static CardViewHolder inflateViewHolder(ViewGroup parent) {
        View card = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.card_events_item, parent, false);
        return new EventsAdapter.EventViewHolder(card);
    }

    @Override
    public void updateViewHolder(@NonNull RecyclerView.ViewHolder viewHolder) {
        super.updateViewHolder(viewHolder);

        if (viewHolder instanceof EventsAdapter.EventViewHolder) {
            EventsAdapter.EventViewHolder holder = (EventsAdapter.EventViewHolder) viewHolder;
            holder.bind(event);
        }
    }

    public void setEvent(Event event) {
        this.event = event;
    }

    @Override
    public Intent getIntent() {
        Intent intent = new Intent(getContext(), EventDetailsActivity.class);
        intent.putExtra("event", event);
        return intent;
    }

    @Override
    protected void discard(@NotNull SharedPreferences.Editor editor) {
        // TODO: add a dismissed attribute to the event table later
        //       see NewsCard for reference
        //       Do nothing for now...
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/ticket/EventCard.kt
package de.tum.`in`.tumcampusapp.component.ui.ticket

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.other.navigation.NavDestination
import de.tum.`in`.tumcampusapp.component.ui.overview.CardInteractionListener
import de.tum.`in`.tumcampusapp.component.ui.overview.CardManager
import de.tum.`in`.tumcampusapp.component.ui.overview.card.Card
import de.tum.`in`.tumcampusapp.component.ui.overview.card.CardViewHolder
import de.tum.`in`.tumcampusapp.component.ui.ticket.adapter.EventsAdapter
import de.tum.`in`.tumcampusapp.component.ui.ticket.model.Event
import de.tum.`in`.tumcampusapp.component.ui.tufilm.KinoActivity
import de.tum.`in`.tumcampusapp.utils.Const

class EventCard(context: Context) : Card(CardManager.CardTypes.EVENT, context) {

    var event: Event? = null

    // TODO(thellmund) Inject this
    private val eventCardsProvider = EventCardsProvider()

    override fun getNavigationDestination(): NavDestination {
        val args = Bundle().apply { putInt(Const.KINO_ID, 1) }
        return NavDestination.Activity(KinoActivity::class.java, args)
    }

    override fun shouldShow(prefs: SharedPreferences): Boolean {
        return event?.dismissed == 0
    }

    override fun discard(editor: SharedPreferences.Editor) {
        event?.let {
            eventCardsProvider.setDismissed(it.id)
        }
    }

    companion object {

        @JvmStatic
        fun inflateViewHolder(
            parent: ViewGroup,
            interactionListener: CardInteractionListener
        ): CardViewHolder {
            val card = LayoutInflater.from(parent.context).inflate(R.layout.card_events_item, parent, false)
            return EventsAdapter.EventViewHolder(card, interactionListener, true)
        }
    }
}
